vowpalwabbit/core/src/reductions/autolink.cc

```diff
@@ -82,17 +82,30 @@ void predict_or_learn(autolink& b, VW::LEARNER::single_learner& base, VW::exampl
   else { b.predict(base, ec); }
 }
 
-VW::LEARNER::base_learner* VW::reductions::autolink_setup(VW::setup_base_i& stack_builder)
+struct options_autolink_v1
 {
-  options_i& options = *stack_builder.get_options();
-  VW::workspace& all = *stack_builder.get_all_pointer();
   uint32_t d;
+};
+
+std::unique_ptr<options_autolink_v1> get_autolink_options_instance(
```

Comment:
What is the contract around this function? Not only does it define the options and parse them it seems it also needs to make the decision if the reduction is enabled or not?

Reply:
Yes it is supposed to define and parse the options, then return the finished configuration, or `nullptr` if the reduction is not enabled